README.md
wc3mods
=======

Warcraft mods for counter strike!

http://wiki.wc3mods.net


war3ft/changelog.txt
v3.0 RC14
 - Bug Fix: [FS#141] Admin command crash fix
 - Bug Fix: Registered cvar twice
 - Bug Fix: [FS#137] - Human Teleport = Restricted Area! (New CVAR and ML Line)
 - Addition: [FS#145] Compatibility with ATAC3.x
 - Addition: Added command war3_version and changed the way the version is handled to support dev builds

v3.0 RC13
 - Bug Fix: [FS#83] Changed on_ResetHud to Ham_Spawn
 - Bug Fix: [FS#132] Chameleon deletes skill selections
 - Bug Fix: [FS#136] bots fix for cz
 - Bug Fix: [FS#133] show player item fix
 - Bug Fix: [FS#134] Cheating on item prices
 - Bug Fix: [FS#92] solution against detecting moles (angle of view)
 - Bug Fix: [FS#131] shadow hunter coconut fix
 - Bug Fix: [FS#123] No serpent wards after respawning. 
 - Change: Remade ADMIN_Log format to more AMXX like
 - Addition: ScreenFade for both attacker and victim of Serpand Wards

v3.0 RC12
 - Bug Fix: [FS#121] Humans can teleport while planting
 - Bug Fix: [FS#81-6] Wrong ultimate charge was being removed (it attempted to remove it from the caster vs. the nearby player with immunity)
 - Bug Fix: [FS#83] Changed respawn method to use Ham
 - Bug Fix: [FS#105] Serpent Ward removal on incorrect event call 

v3.0 RC11b
 - Bug Fix: [FS#109] DoD crash on shopmenu

v3.0 RC11
 - Addition: [FS#92] solution against detecting moles (angle of view)
 - Bug Fix: [FS#97] Removing Human's devotion aura
 - Bug Fix: MySQLX error logging incorrect in some places (reporting wrong query)
 - Bug Fix: [FS#83] Respawn within a player. Respawn dead
 - Bug Fix: [FS#85] Add alive check before kill by SH (patch FS#76)
 - Bug Fix: [FS#88] Always allow chameleon to buy ankh
 - Bug Fix: [FS#95] ITEM_MOLE should disappear when used
 - Bug Fix: Fixed Admin Commands reporting back incorrectly in wc3_admin.log
 - Bug Fix: [FS#96] CT Chameleons with backpack
 - Bug Fix: [FS#90] Locust Swarm didn't attack after finding new target
 - Bug Fix: [FS#89] allow pov-targeting ultimate to work quickly
 - Bug Fix: [FS#102] Chain Lightning actually hits only one enemy
 - Bug Fix: [FS#79] fix invalid player in client_damage()

v3.0 RC10
 - Addition: [FS#68] Playerskills will now print to console if called from console
 - Bug Fix: MYSQLX error function was being called from within SQLITE
 - Bug Fix: [FS#72] Run time error in items.inl
 - Bug Fix: [FS#62] Level 10 sprite wasn't being precached
 - Bug Fix: [FS#60][DOD] Server crash when buying flaming gloves of warmth
 - Bug Fix: Bots casting their ultimate is actually random now, not everytime they attack
 - Bug Fix: If chain lightning was blocked, the wrong person would be notified
 - Bug Fix: [FS#17] Big Bad Voodoo never working
 - Bug Fix: Mole protectant not working
 - Bug Fix: Database spamming problem when a user's ID couldn't be retreived
 - Bug Fix: [FS#76] Hex wasn't disabling all skills
 - Bug Fix: When checking a user's skills the level and skill id were not being matched correctly (in XP_Check)

v3.0 RC9
 - Addition: [FS#29] Client console commands for all items
 - Addition: [FS#59][WEB] Rank based on TOTAL XP - "total" option added to config file
 - Bug Fix: [FS#46] Immolate not doing the proper amount of damage
 - Bug Fix: [FS#47] Warden's Hardened skin was not working correctly
 - Bug Fix: [FS#48] Fixed undead's ultimate going off when hexed
 - Bug Fix: [FS#66] Server crash when using wc3_giveitem with item_id over 17
 - Bug Fix: [FS#63] Server crash on given maps b/c of sprite count
 - Bug Fix: [FS#57] cl_minmodels check failure
 - Bug Fix: [FS#70] g_iSpawnReserved index out of bounds
 - Bug Fix: [FS#67] Mole is shown to spectators when mole is in slot2
 - Bug Fix: Respawn failing with AMXX 1.80
 - Bug Fix: Menus not working with AMXX 1.80
 - Bug Fix: [FS#65] Stuck protection triggered after new round started
 - Bug Fix: [FS#55] Command /levitation now available for any race with skill levitation
 - Bug Fix: [FS#54] Sock item always set gravity to 0.5 (wc3_sock value ignored)
 - Bug Fix: [FS#53] Deselected skills are no longer saved to database (will be saved as skill level as 0)
 - Bug Fix: [FS#35] Player couldn't respawn after WC3_Kill (generally after warden ultimate)
 - Bug Fix: [FS#26] If warden is killed last and immunity check should not occur (i.e. killer shouldn't lose a necklace charge)
 - Bug Fix: [WEB] Searching based on steam id or player name wasn't working

v3.0 RC8
 - Bug Fix: [FS#25] Warden's Shadow Strike had wrong percentage at level 1
 - Bug Fix: [FS#31] Reset item charges when a player joins the server
 - Bug Fix: [FS#28] rings console command not going through proper checks
 - Bug Fix: [FS#32] Giving XP to terrorists in admin menu would not work
 - Bug Fix: [FS#21] Friendly fire no longer affects charges for helm of excellence
 - Bug Fix: [FS#37] Undead ultimate run time error
 - Bug Fix: [FS#38] Grenade trail created on orc's death when grenade drops
 - Bug Fix: [FS#40] Assist XP awarded twice if a user would respawn

v3.0 RC7
 - Addition: wc3_ignore_bots CVAR which works with wc3_min_players (thanks Krazy)
 - Addition: It is now possible do deal damage against a SH w/Big Bad Voodoo if you have immunity
 - Addition: Log file for war3ft exclusive use
 - Addition: (B2) [FS#7] When you respawn with weapons your silence/burst mode will be remembered and set
 - Addition: (B2) wc3_start_level CVAR - You can now start a player's race off at a certain level when they join the server
 - Bug Fix: Gloves not always working correctly
 - Bug Fix: XP Loss at map change
 - Bug Fix: (B2) cl_minmodels check failing at certain times (thanks DesasterUK)
 - Bug Fix: (B2) Conversion issue when war3users table didn't exist (thanks Coksnuss)
 - Bug Fix: (B2) Getting stuck in a wall from Banish or failed Teleport - you should now be teleported back to your spawn (if possible) instead of left stuck
 - Bug Fix: (B2) Big Bad Voodoo works now - opps
 - Bug Fix: (B2) Ankh/Mole will no longer work if you replace the item
 - Bug Fix: (B2) Saving by IP will not strip the port from the address

war3ft/war3ft/XP.h
/*
*	XP Header File
*/

// Objective Modifiers
#define DEFUSING_BOMB		10			// XP awarded when the user starts to defuse the bomb
#define DEFUSED_BOMB		20			// XP awarded when the user defuses the bomb
#define PLANTING_BOMB		10			// XP awarded when the user starts planting the bomb
#define PLANT_BOMB			20			// XP awarded when the user plants the bomb
#define SPAWN_BOMB			10			// XP awarded when the user spawns with the bomb
#define BOMB_PICKUP			10			// XP awarded when the user picks up the bomb
#define TOUCH_HOSTAGE		10			// XP awarded when the user touches a hostage
#define RESCUE_HOSTAGE		20			// XP awarded when the user rescues the hostage
#define KILL_HOSTAGE		10			// XP lost when killing a hostage
#define SPAWN_VIP			10			// XP awarded for spawning as the VIP
#define ESCAPE_VIP			20			// XP awarded for escaping as the VIP
#define OBJ_RADIUS			500			// Nearby radius to award XP for helping complete objectives

// Kill modifiers
#define KILL_HEADSHOT		10			// XP awarded for getting a headshot
#define KILL_HOSTAGE_SAVER	10			// XP awarded for killing the hostage saver
#define	KILL_DEFUSER		10			// XP awarded for killing the defuser
#define KILL_PLANTER		10			// XP awarded for killing the planter
#define KILL_BOMB_CARRIER	10			// XP awarded for killing the bomb carrier
#define KILL_VIP			20			// XP awarded for killing the VIP
#define KILL_RADIUS			250			// Nearby radius to award XP

#define WIN_ROUND			20			// XP awarded for winning the round


// Holds information about the player
enum
{
	PLR_BOMB_DEFUSER = 1,
	PLR_BOMB_PLANTER,
	PLR_HOSTAGE_RESCUER,
	PLR_VIP,
	PLR_BOMB_CARRIER,
};

new g_iPlayerRole[33];
new bool:bHasBegunPlantingOrDefusing[33];

// Holds the XP Multipliers per weapon
new Float:fWpnXPMultiplier[CSW_WAR3_MAX+1] =		{1.0};

// Amount of XP needed to gain a level
new iXPLevelShortTerm[11]				= {0,150,300,600,1000,1500,2100,2800,3400,4500,5500};
new iXPLevelSaved[11]					= {0,100,200,400,800,1600,3200,6400,12800,25600,51200};

// Amount of XP awarded when killing a user of this level
new iXPGivenShortTerm[11]				= {10,15,25,35,40,50,60,70,80,90,95};
new iXPGivenSaved[11]					= {6,8,10,12,14,16,18,20,24,28,32};